always-hydrate.md

```diff
@@ -0,0 +1,133 @@
+# Design for always-hydrate feature
+
+## Goal
+The goal of this feature is to enable a developer to communicate that they care about a set of paths,
+and want their IO in those paths to always be fast. E.g. they will always build a certain subfolder,
+and they don't want to pay for file-by-file downloads within that folder.
+
+Today, a user can partially communicate this desire with the `gvfs prefetch --files` or `gvfs
+prefetch --folders` commands to bulk download the files they care about. And they can further specify
+the `--hydrate` flag to either command to bulk hydrate those files into the working directory.
+
+However, this is not a great experience because it's just a one-time prefetch and hydrate. What that
+means is that the next time the projection changes, e.g. after a `git checkout`, anywhere from some
+to all of those files can become dehydrated, and the user will again pay the cost of file-by-file
+downloads while trying to build.
+
+With this feature, we want the developer to make one gesture and communicate that a set of files
+should always be kept hydrated, even after the projection changes. That being said, a secondary goal
+is to not add any significant time to commands like `git checkout` that change the projection, so the
+implementation should ensure that these are done as background downloads that do not block these
+commands.
+
+## UX
+This is really an extension of the existing prefetch functionality. As such, it makes sense to tie
+the UX to the existing `gvfs prefetch` verb.
+
+Side note on the existing prefetch verb: it's really two different commands in one, and it's
+unfortunate that they are combined. The command `gvfs prefetch --commits` is entirely different in
+its goals, usage, and implementation than either of `gvfs prefetch --files` or `gvfs prefetch 
+--folders` (and the latter two can be combined with each other, but not with `--commits`). Ideally we
+would have had `gvfs prefetch-commits` and `gvfs prefetch-files` as two separate commands, which
+would have then led to a cleaner design for this new addition. I'm open to other suggestions here
+if anyone sees a better pattern for extending the current verbs.
+
+### Proposed UX for always-hydrate commands:
+Adding new entries: 
+
+Add an `--always-hydrate` flag to the existing `--files` and `--folders` variants of `gvfs prefetch`.
+This will communicate that whatever list of folders, files, and wildcards that were passed in to the
+command should be saved and reapplied after every projection change
+
+Syntax:
+
+```
+    gvfs prefetch 
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      [--folders-list <path of file containing newline-delimited paths of folders, no wildcards>] 
+      [--[always-]hydrate]
+```
+
+The only change from today's syntax is the optional `[always-]` on the hydrate flag. 
```

Comment:
This implies that we support a mixed set of folders and files in the `always-hydrate` set. Is that the correct level of granularity? Should we start with only folders, since it seems quite fiddly ask a user to enumerate interesting files for us? Is there a scenario I'm not considering?

Reply:
The scenario where it's useful is `--files *.csproj`, or any other build definition files